src/deepsparse/pipeline.py

```diff
@@ -0,0 +1,375 @@
+            Pass None for the default
+        :param input_shapes: list of shapes to set ONNX the inputs to. Pass None
+            to use model as-is. Default is None
+        :param kwargs: extra task specific kwargs to be passed to task Pipeline
+            implementation
+        :param alias: optional name to give this pipeline instance, useful when
+            inferencing with multiple models. Default is None
+        :return: pipeline object initialized for the given task
+        """
+        task = task.lower().replace("-", "_")
+
+        # extra step to register pipelines for a given task domain
+        # for cases where imports should only happen once a user specifies
+        # that domain is to be used. (ie deepsparse.transformers will auto
+        # install extra packages so should only import and register once a
+        # transformers task is specified)
+        SupportedTasks.check_register_task(task)
+
+        if task not in _REGISTERED_PIPELINES:
+            raise ValueError(
+                f"Unknown Pipeline task {task}. Pipeline tasks should be "
+                "must be declared with the Pipeline.register decorator. Currently "
+                f"registered pipelines: {list(_REGISTERED_PIPELINES.keys())}"
+            )
+
+        return _REGISTERED_PIPELINES[task](
+            model_path=model_path,
+            engine_type=engine_type,
+            batch_size=batch_size,
+            num_cores=num_cores,
+            scheduler=scheduler,
+            input_shapes=input_shapes,
+            alias=alias,
+            **kwargs,
+        )
+
+    @classmethod
+    def register(cls, task: str, task_aliases: Optional[List[str]]):
+        """
+        Pipeline implementer class decorator that registers the pipeline
+        task name and its aliases as valid tasks that can be used to load
+        the pipeline through `Pipeline.create()`.
+
+        Multiple pipelines may not have the same task name. An error will
+        be raised if two different pipelines attempt to register the same task name
+
+        :param task: main task name of this pipeline
+        :param task_aliases: list of extra task names that may be used to reference
+            this pipeline
+        """
+        task_names = [task]
+        if task_aliases:
+            task_names.extend(task_aliases)
+
+        def _register_task(task_name, pipeline_class):
+            if task_name in _REGISTERED_PIPELINES and (
+                pipeline_class is not _REGISTERED_PIPELINES[task_name]
+            ):
+                raise RuntimeError(
+                    f"task {task_name} already registered by Pipeline.register. "
+                    f"attempting to register pipeline: {pipeline_class}, but"
+                    f"pipeline: {_REGISTERED_PIPELINES[task_name]}, already registered"
+                )
+            _REGISTERED_PIPELINES[task_name] = pipeline_class
+
+        def decorator(pipeline_class: Pipeline):
```

Comment:
Could we wrap this in `functools.wraps(cls, updated=())`? to ensure proper propagation of dunder attributes

Reply:
`cls` would refer to `Pipeline` here, not the implementing `pipeline_class`,  I think as long as I fix the decorator to `return pipeline_class` the wrapping should go through fine

Reply:
Changes look good, resolving this thread!